File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()  # dummy node
        tail = head

        while list1 or list2:
            if not list1:
                tail.next = list2
                break
            if not list2:
                tail.next = list1
                break
            if list1.val < list2.val:
                nd = list1
                list1 = list1.next
            else:
                nd = list2
                list2 = list2.next
            tail.next = nd
            tail = tail.next
        return head.next

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode lhs, ListNode rhs) {
        var dummy = new ListNode();
        var tail = dummy;
        
        while (lhs != null && rhs != null) {
            if(lhs.val < rhs.val) {
                tail.next = lhs;
                lhs = lhs.next;
            } else {
                tail.next = rhs;
                rhs = rhs.next;
            }
            tail = tail.next;
        }
        if (lhs != null) {
            tail.next = lhs;
        } else {
            tail.next = rhs;
        }

        return dummy.next;
    }
}